File: frontend/src/types/design.ts
// Mirrors backend/internal/design/design.go field-for-field (json tags
// match). Keep these two in sync by hand — there is no shared codegen
// between the Go module and the frontend workspace.

export type LayoutMode = 'one' | 'two' | 'mix'

export interface SectionRef {
  sectionType: string
  customSectionId: string | null
  isVisible: boolean
  titleOverride: string | null
}

export interface SectionBand {
  full?: SectionRef[]
  left?: SectionRef[]
  right?: SectionRef[]
}

export interface SectionOrder {
  one: { sections: SectionRef[] }
  two: { left: SectionRef[]; right: SectionRef[] }
  mix: { bands: SectionBand[] }
}

export interface Layout {
  mode: LayoutMode
  columnWidths: { left: number; right: number }
}

export interface Page {
  format: 'A4' | 'Letter'
  marginTop: number
  marginBottom: number
  marginLeft: number
  marginRight: number
}

export interface Typography {
  fontFamily: string
  /** "inherit" falls back to fontFamily, or one of the same font options. */
  nameFontFamily: string
  baseFontSizePt: number
  lineHeight: number
  nameFontSizePt: number
  headlineFontSizePt: number
  sectionHeadingFontSizePt: number
  entryHeaderFontSizePt: number
}

export interface ApplyAccent {
  name: boolean
  headings: boolean
  dates: boolean
  icons: boolean
}

export interface Colors {
  text: string
  accent: string
  background: string
  applyAccent: ApplyAccent
}

export interface Heading {
  style: 'line' | 'box' | 'underline' | 'thinLine' | 'simple'
  capitalization: 'none' | 'uppercase'
}

export interface HeaderPhoto {
  show: boolean
  size: 's' | 'm' | 'l'
}

export interface Header {
  photo: HeaderPhoto
  alignText: 'left' | 'center'
  jobTitlePosition: 'below' | 'inline'
}

export interface EntryLayout {
  dateDisplayMode: 'right' | 'left' | 'belowTitle'
  subtitleStyle: 'normal' | 'italic'
}

export type SectionDisplayMode = 'grid' | 'text' | 'bullets'

export interface SectionDisplay {
  skills: SectionDisplayMode
  languages: SectionDisplayMode
  certifications: SectionDisplayMode
}

export interface Spacing {
  sectionGap: number
  entryGap: number
  bulletGap: number
}

// Named distinctly from "Links" to avoid colliding with the unrelated
// concept of a resume's own header links (LinkedIn/GitHub URLs), which live
// in content (ContentHeader.links), not design.
export interface LinkStyle {
  showIcon: boolean
  underline: boolean
  useAccentColor: boolean
}

export interface Footer {
  showPageNumbers: boolean
  showEmail: boolean
  showName: boolean
}

export interface ResumeDesign {
  templateId: string
  layout: Layout
  sectionOrder: SectionOrder
  page: Page
  typography: Typography
  colors: Colors
  heading: Heading
  header: Header
  entryLayout: EntryLayout
  sectionDisplay: SectionDisplay
  spacing: Spacing
  linkStyle: LinkStyle
  footer: Footer
  dateFormat: 'MM/YYYY' | 'Mon YYYY' | 'YYYY'
}

File: backend/internal/design/design.go
package design

type LayoutMode string

const (
	LayoutOne LayoutMode = "one"
	LayoutTwo LayoutMode = "two"
	LayoutMix LayoutMode = "mix"
)

type SectionRef struct {
	SectionType     string  `json:"sectionType"`
	CustomSectionID *string `json:"customSectionId"`
	IsVisible       bool    `json:"isVisible"`
	TitleOverride   *string `json:"titleOverride"`
}

type SectionBand struct {
	Full  []SectionRef `json:"full,omitempty"`
	Left  []SectionRef `json:"left,omitempty"`
	Right []SectionRef `json:"right,omitempty"`
}

type SectionOrder struct {
	One struct {
		Sections []SectionRef `json:"sections"`
	} `json:"one"`
	Two struct {
		Left  []SectionRef `json:"left"`
		Right []SectionRef `json:"right"`
	} `json:"two"`
	Mix struct {
		Bands []SectionBand `json:"bands"`
	} `json:"mix"`
}

type Layout struct {
	Mode         LayoutMode `json:"mode"`
	ColumnWidths struct {
		Left  int `json:"left"`
		Right int `json:"right"`
	} `json:"columnWidths"`
}

type Page struct {
	Format       string `json:"format"`
	MarginTop    int    `json:"marginTop"`
	MarginBottom int    `json:"marginBottom"`
	MarginLeft   int    `json:"marginLeft"`
	MarginRight  int    `json:"marginRight"`
}

type Typography struct {
	FontFamily               string  `json:"fontFamily"`
	NameFontFamily           string  `json:"nameFontFamily"`
	BaseFontSizePt           float64 `json:"baseFontSizePt"`
	LineHeight               float64 `json:"lineHeight"`
	NameFontSizePt           float64 `json:"nameFontSizePt"`
	HeadlineFontSizePt       float64 `json:"headlineFontSizePt"`
	SectionHeadingFontSizePt float64 `json:"sectionHeadingFontSizePt"`
	EntryHeaderFontSizePt    float64 `json:"entryHeaderFontSizePt"`
}

type ApplyAccent struct {
	Name     bool `json:"name"`
	Headings bool `json:"headings"`
	Dates    bool `json:"dates"`
	Icons    bool `json:"icons"`
}

type Colors struct {
	Text        string      `json:"text"`
	Accent      string      `json:"accent"`
	Background  string      `json:"background"`
	ApplyAccent ApplyAccent `json:"applyAccent"`
}

type Heading struct {
	Style          string `json:"style"`
	Capitalization string `json:"capitalization"`
}

type HeaderPhoto struct {
	Show bool   `json:"show"`
	Size string `json:"size"`
}

type Header struct {
	Photo            HeaderPhoto `json:"photo"`
	AlignText        string      `json:"alignText"`
	JobTitlePosition string      `json:"jobTitlePosition"`
}

type EntryLayout struct {
	DateDisplayMode string `json:"dateDisplayMode"`
	SubtitleStyle   string `json:"subtitleStyle"`
}

type SectionDisplay struct {
	Skills         string `json:"skills"`
	Languages      string `json:"languages"`
	Certifications string `json:"certifications"`
}

type Spacing struct {
	SectionGap int `json:"sectionGap"`
	EntryGap   int `json:"entryGap"`
	BulletGap  int `json:"bulletGap"`
}

type LinkStyle struct {
	ShowIcon       bool `json:"showIcon"`
	Underline      bool `json:"underline"`
	UseAccentColor bool `json:"useAccentColor"`
}

type Footer struct {
	ShowPageNumbers bool `json:"showPageNumbers"`
	ShowEmail       bool `json:"showEmail"`
	ShowName        bool `json:"showName"`
}

type ResumeDesign struct {
	TemplateID     string         `json:"templateId"`
	Layout         Layout         `json:"layout"`
	SectionOrder   SectionOrder   `json:"sectionOrder"`
	Page           Page           `json:"page"`
	Typography     Typography     `json:"typography"`
	Colors         Colors         `json:"colors"`
	Heading        Heading        `json:"heading"`
	Header         Header         `json:"header"`
	EntryLayout    EntryLayout    `json:"entryLayout"`
	SectionDisplay SectionDisplay `json:"sectionDisplay"`
	Spacing        Spacing        `json:"spacing"`
	LinkStyle      LinkStyle      `json:"linkStyle"`
	Footer         Footer         `json:"footer"`
	DateFormat     string         `json:"dateFormat"`
}
